File: draw3D.py
import vtk

class CustomInteractorStyle(vtk.vtkInteractorStyleTrackballCamera):
    def __init__(self, compositeTransform, startPos, hudAxes, camera, camera_initial):
        super(CustomInteractorStyle, self).__init__()
        self.compositeTransform = compositeTransform
        self.startPos = startPos
        self.hudAxes = hudAxes
        self.camera = camera
        self.camera_initial = camera_initial
        # Add an observer for key press events.
        self.AddObserver("KeyPressEvent", self.on_key_press)

    def updateHUDAxes(self):
        mat = vtk.vtkMatrix4x4()
        self.compositeTransform.GetMatrix(mat)
        # Zero out translation components (we want only rotation in the HUD).
        mat.SetElement(0, 3, 0.0)
        mat.SetElement(1, 3, 0.0)
        mat.SetElement(2, 3, 0.0)
        self.hudAxes.SetUserMatrix(mat)

    def on_key_press(self, obj, event):
        key = self.GetInteractor().GetKeySym()
        angleStep = 5.0  # degrees per key press
        rotated = False

        if key == "Up":
            self.compositeTransform.RotateX(angleStep)
            # print("Rotated composite around X by", angleStep)
            rotated = True
        elif key == "Down":
            self.compositeTransform.RotateX(-angleStep)
            # print("Rotated composite around X by", -angleStep)
            rotated = True
        elif key == "Left":
            self.compositeTransform.RotateY(angleStep)
            # print("Rotated composite around Y by", angleStep)
            rotated = True
        elif key == "Right":
            self.compositeTransform.RotateY(-angleStep)
            # print("Rotated composite around Y by", -angleStep)
            rotated = True
        elif key == "space":
            # Reset composite transform.
            self.compositeTransform.Identity()
            self.compositeTransform.Translate(*self.startPos)
            # Also restore the camera's initial parameters.
            self.camera.SetPosition(*self.camera_initial['position'])
            self.camera.SetFocalPoint(*self.camera_initial['focal_point'])
            self.camera.SetViewUp(*self.camera_initial['view_up'])
            # Reset the camera clipping range.
            renderer = self.GetInteractor().GetRenderWindow().GetRenderers().GetFirstRenderer()
            renderer.ResetCameraClippingRange()
            # print("Reset composite transform and camera to initial state.")
            rotated = True
        else:
            super(CustomInteractorStyle, self).OnKeyPress()

        if rotated:
            self.updateHUDAxes()
            self.GetInteractor().GetRenderWindow().Render()

def read_vertices(filename):
    vertices = []
    with open(filename, "r") as f:
        for line in f:
            parts = line.strip().split()
            if not parts:
                continue
            if len(parts) == 2:
                x, y = map(float, parts)
                z = 0.0
            else:
                x, y, z = map(float, parts[:3])
            vertices.append((x, y, z))
    return vertices

def read_edges(filename):
    edges = []
    with open(filename, "r") as f:
        for line in f:
            parts = line.strip().split()
            if len(parts) < 2:
                continue
            u, v = map(int, parts[:2])
            edges.append((u, v))
    return edges

def main():
    # File paths (adjust as needed)
    vertex_file = "./tmp/graph/embedding.txt"
    edge_file   = "graph.txt"

    # Read data.
    vertices = read_vertices(vertex_file)
    edges = read_edges(edge_file)

    print("Loaded", len(vertices), "vertices")
    print("Loaded", len(edges), "edges")

    if not vertices:
        print("No vertices loaded!")
        return

    # Create vtkPoints and add vertices.
    vtk_points = vtk.vtkPoints()
    for pt in vertices:
        vtk_points.InsertNextPoint(pt)

    # Create a vtkCellArray for the edges.
    vtk_lines = vtk.vtkCellArray()
    for (u, v) in edges:
        if u < 0 or v < 0 or u >= len(vertices) or v >= len(vertices):
            continue
        line = vtk.vtkLine()
        line.GetPointIds().SetId(0, u)
        line.GetPointIds().SetId(1, v)
        vtk_lines.InsertNextCell(line)

    # Create polydata to hold the graph.
    polyData = vtk.vtkPolyData()
    polyData.SetPoints(vtk_points)
    polyData.SetLines(vtk_lines)

    # Create a mapper and actor for the graph.
    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputData(polyData)

    actor = vtk.vtkActor()
    actor.SetMapper(mapper)
    actor.GetProperty().SetColor(65/255.0, 105/255.0, 225/255.0)  # royal blue
    actor.GetProperty().SetLineWidth(2)

    # Create a composite transform and apply it only to the graph actor.
    compositeTransform = vtk.vtkTransform()
    startPos = (0.0, 0.0, 0.0)
    compositeTransform.Translate(*startPos)
    actor.SetUserTransform(compositeTransform)

    # Create renderer, render window, and interactor.
    renderer = vtk.vtkRenderer()
    # Set background to a white-grey color.
    renderer.SetBackground(0.95, 0.95, 0.95)

    renderWindow = vtk.vtkRenderWindow()
    renderWindow.AddRenderer(renderer)

    interactor = vtk.vtkRenderWindowInteractor()
    interactor.SetRenderWindow(renderWindow)

    # Add the graph actor.
    renderer.AddActor(actor)

    # Create a separate axes actor for the orientation marker (HUD).
    hudAxes = vtk.vtkAxesActor()

    # Reset the camera to include all actors.
    renderer.ResetCamera()
    renderWindow.Render()

    # Get the active camera and store its initial parameters.
    camera = renderer.GetActiveCamera()
    camera_initial = {
        'position': camera.GetPosition(),
        'focal_point': camera.GetFocalPoint(),
        'view_up': camera.GetViewUp()
    }

    # Instantiate the custom interactor style with the composite transform, hudAxes, and camera.
    style = CustomInteractorStyle(compositeTransform, startPos, hudAxes, camera, camera_initial)
    interactor.SetInteractorStyle(style)

    # Add an orientation marker widget to display the HUD axes.
    orientationWidget = vtk.vtkOrientationMarkerWidget()
    orientationWidget.SetOrientationMarker(hudAxes)
    orientationWidget.SetInteractor(interactor)
    orientationWidget.SetViewport(0.0, 0.0, 0.2, 0.2)
    orientationWidget.EnabledOn()
    orientationWidget.InteractiveOff()

    renderWindow.SetSize(800, 600)
    renderWindow.SetWindowName("3D Graph with Rotating Coordinates (VTK - Python)")

    objExporter = vtk.vtkOBJExporter()
    objExporter.SetFilePrefix("graph")  # This will create "graph.obj" and "graph.mtl"
    objExporter.SetRenderWindow(renderWindow)
    objExporter.Write()

    interactor.Start()

if __name__ == "__main__":
    main()

File: draw3D.cpp
#include <vtkSmartPointer.h>
#include <vtkInteractorStyleTrackballCamera.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderWindow.h>
#include <vtkRenderer.h>
#include <vtkPoints.h>
#include <vtkCellArray.h>
#include <vtkPolyData.h>
#include <vtkPolyDataMapper.h>
#include <vtkActor.h>
#include <vtkProperty.h>
#include <vtkTransform.h>
#include <vtkMatrix4x4.h>
#include <vtkAxesActor.h>
#include <vtkOrientationMarkerWidget.h>
#include <vtkCamera.h>
#include <vtkLine.h>
#include <vtkOBJExporter.h>
#include <vtkNew.h>
#include <vtkCommand.h>
#include <vtkObjectFactory.h>
#include <vtkRendererCollection.h>

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <array>
#include <string>


class CustomInteractorStyle : public vtkInteractorStyleTrackballCamera {
public:
  static CustomInteractorStyle* New();
  vtkTypeMacro(CustomInteractorStyle, vtkInteractorStyleTrackballCamera);

  CustomInteractorStyle() {
    this->CompositeTransform = nullptr;
    this->HUDAxes = nullptr;
    this->Camera = nullptr;
    for (int i = 0; i < 3; i++) {
      this->StartPos[i] = 0.0;
      this->CameraInitialPosition[i] = 0.0;
      this->CameraInitialFocal[i] = 0.0;
      this->CameraInitialViewUp[i] = 0.0;
    }
  }

  void SetCompositeTransform(vtkTransform* transform) {
    this->CompositeTransform = transform;
  }

  void SetStartPos(double pos[3]) {
    for (int i = 0; i < 3; ++i)
      this->StartPos[i] = pos[i];
  }

  void SetHUDAxes(vtkAxesActor* axes) {
    this->HUDAxes = axes;
  }

  void SetCamera(vtkCamera* cam) {
    this->Camera = cam;
  }

  void SetCameraInitial(double pos[3], double focal[3], double viewup[3]) {
    for (int i = 0; i < 3; ++i) {
      this->CameraInitialPosition[i] = pos[i];
      this->CameraInitialFocal[i] = focal[i];
      this->CameraInitialViewUp[i] = viewup[i];
    }
  }

  virtual void OnKeyPress() override {
    vtkRenderWindowInteractor* interactor = this->GetInteractor();
    std::string key = interactor->GetKeySym();
    double angleStep = 5.0;
    bool rotated = false;

    if (key == "Up") {
      this->CompositeTransform->RotateX(angleStep);
      rotated = true;
    }
    else if (key == "Down") {
      this->CompositeTransform->RotateX(-angleStep);
      rotated = true;
    }
    else if (key == "Left") {
      this->CompositeTransform->RotateY(angleStep);
      rotated = true;
    }
    else if (key == "Right") {
      this->CompositeTransform->RotateY(-angleStep);
      rotated = true;
    }
    else if (key == "space") {

      this->CompositeTransform->Identity();
      this->CompositeTransform->Translate(this->StartPos[0], this->StartPos[1], this->StartPos[2]);

      if (this->Camera) {
        this->Camera->SetPosition(this->CameraInitialPosition);
        this->Camera->SetFocalPoint(this->CameraInitialFocal);
        this->Camera->SetViewUp(this->CameraInitialViewUp);
      }
      vtkRenderer* renderer = interactor->GetRenderWindow()->GetRenderers()->GetFirstRenderer();
      if (renderer) {
        renderer->ResetCameraClippingRange();
      }
      rotated = true;
    }
    else {

      vtkInteractorStyleTrackballCamera::OnKeyPress();
    }

    if (rotated) {
      this->UpdateHUDAxes();
      interactor->GetRenderWindow()->Render();
    }
  }

protected:
  void UpdateHUDAxes() {
    vtkNew<vtkMatrix4x4> mat;
    this->CompositeTransform->GetMatrix(mat);

    mat->SetElement(0, 3, 0.0);
    mat->SetElement(1, 3, 0.0);
    mat->SetElement(2, 3, 0.0);
    if (this->HUDAxes) {
      this->HUDAxes->SetUserMatrix(mat);
    }
  }

  vtkTransform* CompositeTransform;
  double StartPos[3];
  vtkAxesActor* HUDAxes;
  vtkCamera* Camera;
  double CameraInitialPosition[3];
  double CameraInitialFocal[3];
  double CameraInitialViewUp[3];
};

vtkStandardNewMacro(CustomInteractorStyle);


std::vector<std::array<double, 3>> read_vertices(const std::string& filename) {
    std::vector<std::array<double, 3>> vertices;
    std::ifstream infile(filename);
    if (!infile) {
        std::cerr << "Could not open vertex file: " << filename << std::endl;
        return vertices;
    }
    std::string line;
    while (std::getline(infile, line)) {
        if (line.empty()) continue;
        std::istringstream iss(line);
        std::vector<double> vals;
        double v;
        while (iss >> v) {
            vals.push_back(v);
        }
        if (vals.size() == 2) {
            vertices.push_back({vals[0], vals[1], 0.0});
        }
        else if (vals.size() >= 3) {
            vertices.push_back({vals[0], vals[1], vals[2]});
        }
    }
    return vertices;
}


std::vector<std::pair<int, int>> read_edges(const std::string& filename) {
    std::vector<std::pair<int, int>> edges;
    std::ifstream infile(filename);
    if (!infile) {
        std::cerr << "Could not open edge file: " << filename << std::endl;
        return edges;
    }
    std::string line;
    while (std::getline(infile, line)) {
        if (line.empty()) continue;
        std::istringstream iss(line);
        int u, v;
        if (iss >> u >> v) {
            edges.push_back(std::make_pair(u, v));
        }
    }
    return edges;
}

int main(int argc, char* argv[]) {

    std::string vertex_file = "./tmp/graph/embedding.txt";
    std::string edge_file   = "graph.txt";


    auto vertices = read_vertices(vertex_file);
    auto edges = read_edges(edge_file);

    std::cout << "Loaded " << vertices.size() << " vertices" << std::endl;
    std::cout << "Loaded " << edges.size() << " edges" << std::endl;

    if (vertices.empty()) {
        std::cerr << "No vertices loaded!" << std::endl;
        return EXIT_FAILURE;
    }


    vtkNew<vtkPoints> vtk_points;
    for (const auto& pt : vertices) {
        vtk_points->InsertNextPoint(pt[0], pt[1], pt[2]);
    }


    vtkNew<vtkCellArray> vtk_lines;
    for (const auto& edge : edges) {
        int u = edge.first;
        int v = edge.second;
        if (u < 0 || v < 0 || u >= static_cast<int>(vertices.size()) || v >= static_cast<int>(vertices.size()))
            continue;
        vtkNew<vtkLine> line;
        line->GetPointIds()->SetId(0, u);
        line->GetPointIds()->SetId(1, v);
        vtk_lines->InsertNextCell(line);
    }


    vtkNew<vtkPolyData> polyData;
    polyData->SetPoints(vtk_points);
    polyData->SetLines(vtk_lines);


    vtkNew<vtkPolyDataMapper> mapper;
    mapper->SetInputData(polyData);

    vtkNew<vtkActor> actor;
    actor->SetMapper(mapper);
    actor->GetProperty()->SetColor(65.0/255.0, 105.0/255.0, 225.0/255.0); // royal blue
    actor->GetProperty()->SetLineWidth(2);

    vtkNew<vtkTransform> compositeTransform;
    double startPos[3] = {0.0, 0.0, 0.0};
    compositeTransform->Translate(startPos);
    actor->SetUserTransform(compositeTransform);


    vtkNew<vtkRenderer> renderer;
    renderer->SetBackground(0.95, 0.95, 0.95);

    vtkNew<vtkRenderWindow> renderWindow;
    renderWindow->AddRenderer(renderer);

    vtkNew<vtkRenderWindowInteractor> interactor;
    interactor->SetRenderWindow(renderWindow);


    renderer->AddActor(actor);


    vtkNew<vtkAxesActor> hudAxes;


    renderer->ResetCamera();
    renderWindow->Render();


    vtkCamera* camera = renderer->GetActiveCamera();
    double cameraInitialPosition[3];
    double cameraInitialFocal[3];
    double cameraInitialViewUp[3];
    camera->GetPosition(cameraInitialPosition);
    camera->GetFocalPoint(cameraInitialFocal);
    camera->GetViewUp(cameraInitialViewUp);


    vtkNew<CustomInteractorStyle> style;
    style->SetCompositeTransform(compositeTransform);
    style->SetStartPos(startPos);
    style->SetHUDAxes(hudAxes);
    style->SetCamera(camera);
    style->SetCameraInitial(cameraInitialPosition, cameraInitialFocal, cameraInitialViewUp);
    interactor->SetInteractorStyle(style);


    vtkNew<vtkOrientationMarkerWidget> orientationWidget;
    orientationWidget->SetOrientationMarker(hudAxes);
    orientationWidget->SetInteractor(interactor);
    orientationWidget->SetViewport(0.0, 0.0, 0.2, 0.2);
    orientationWidget->EnabledOn();
    orientationWidget->InteractiveOff();

    renderWindow->SetSize(800, 600);
    renderWindow->SetWindowName("3D Graph with Rotating Coordinates (VTK - C++)");


    vtkNew<vtkOBJExporter> objExporter;
    objExporter->SetFilePrefix("graph"); // This will create "graph.obj" and "graph.mtl"
    objExporter->SetRenderWindow(renderWindow);
    objExporter->Write();

    interactor->Start();

    return EXIT_SUCCESS;
}
